stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -808,38 +808,60 @@ function dot(x::AbstractVector, A::AbstractMatrix, y::AbstractVector)
         return dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
     end
 
-    if m == n
-        if istril(A)
-            if istriu(A)
-                return dot(x, Diagonal(A), y)
-            elseif isbanded(A, -1, 0)
-                return dot(x, Bidiagonal(A, 'L'), y)
-            else
-                return dot(x, LowerTriangular(A), y)
+    T = typeof(dot(first(x), first(A), first(y)))
+    s = zero(T)
+    @inbounds for j in 1:n
+        yj = y[j]
+        if !iszero(yj)
+            temp = zero(T)
+            @simd for i in 1:m
+                temp += adjoint(x[i]) * A[i,j]
             end
+            s += temp * yj
         end
-        if istriu(A)
-            if isbanded(A, 0, 1)
-                return dot(x, Bidiagonal(A, 'U'), y)
-            else
-                return dot(x, UpperTriangular(A), y)
+    end
+    return s
+end
+function dot(x::AbstractVector, adjA::Adjoint, y::AbstractVector)
```

Comment:
Can't this just be
```jl
dot(x::AbstractVector, adjA::Adjoint, y::AbstractVector) = adjoint(dot(y, A', x))
```
?

Reply:
Sure. So you want to not treat the non-real transpose case explicitly? I didn't want to handle all other structured cases with Transpose/Adjoint anyway, so at best these cases here. I'm fine with reducing the code as you propose.

Reply:
> So you want to not treat the non-real transpose case explicitly?

I don't think it's worth the code bloat; we can always add an optimized method later if there is a demonstrated need.

Reply:
Can be slightly generalized to:
```jl
dot(x::AbstractVector, transA::Transpose{<:Real}, y::AbstractVector) = adjoint(dot(y, transA.parent, x))
```

Reply:
So I went with your short solutions. Many thanks!